### PR TITLE
feat(core): add help,list and better DX to workspace-schematic

### LINKED ISSUE #2267: @nrwl/workspace - nx workspace-schematic my-schematic --help doesn't invoke help on custom schematic
# Prerequisites

- [x] I am running the latest version
- [x] I checked the documentation (nx.dev) and found no answer
- [x] I checked to make sure that this issue has not already been filed
- [x] I'm reporting the issue to the correct repository (not related to React, Angular or any dependency)

## Expected Behavior

```
nx workspace-schematic my-schematic --help
```

Should print out to terminal help based on custom schematic schema

## Current Behavior

- I've developed custom nx schematic (generated via `nx g @nrwl/workspace:workspace-schematic my-schematic` 
- when I wanna get help (by printing out supported schematic schema options) by executing `nx workspace-schematic my-schematic --help`, instead of getting proper help I'm getting output of `nx workspace-schematic --help`
image

### Context

```
 @nrwl/angular : Not Found
  @nrwl/cli : 8.9.0
  @nrwl/cypress : 8.9.0
  @nrwl/eslint-plugin-nx : 8.9.0
  @nrwl/express : Not Found
  @nrwl/jest : 8.9.0
  @nrwl/linter : 8.9.0
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : 8.9.0
  @nrwl/react : 8.9.0
  @nrwl/schematics : Not Found
  @nrwl/tao : 8.9.0
  @nrwl/web : 8.9.0
  @nrwl/workspace : 8.9.0
  typescript : 3.7.3
```

### PR DESCRIPTION
**Pull request by Hotell**
ISSUES CLOSED: #2267

## Before
2020-08-27 at 5 38 PM

2020-08-27 at 5 39 PM

## After

- better command help description
image

- `-l`, `--list` support with better DX ( description + nicer terminal output)
2020-08-27 at 5 35 PM

- `--help`, `-h` support for called schematic

image

2020-08-27 at 5 34 PM

### Additional Notes

- [x] I wasn't able to get proper `--help` support as yargs will consume default help of command chain (maybe I'm missing something)
- [ ] I'll need to add e2e tests 
  - ~can someone help me out with this ?~ figured on my own
  - ~blocked by https://github.com/nrwl/nx/pull/3621 (where are grouped e2e properly, after that is merged I'll add e2e to prevent conflicts)~

### CODE REVIEW

File: packages/workspace/src/command-line/workspace-schematic.ts [hunk 1]
```diff
@@ -98,29 +155,42 @@ function compileToolsDir(outDir: string) {
...
   };
 }
 
-function saveCollection(dir: string, collection: any) {
-  writeFileSync(
-    path.join(dir, 'workspace-schematics.json'),
-    JSON.stringify(collection)
-  );
+function getSchemaJson(schemaPath: string): SchematicSchema {
+  const schema = readJsonFile<SchematicSchema>(schemaPath);
```
Comment (vsavkin) [inline packages/workspace/src/command-line/workspace-schematic.ts:187]: inline variable

File: packages/workspace/src/command-line/workspace-schematic.ts [hunk 2]
```diff
@@ -139,12 +209,27 @@ function createWorkflow(dryRun: boolean) {
...
+  logger: logging.Logger
+) {
   try {
-    const engineHost = new NodeModulesEngineHost();
-    const engine = new SchematicEngine(engineHost);
-    const collection = engine.createCollection(collectionName);
-    logger.info(engine.listSchematicNames(collection).join('\n'));
+    const bodyLines: string[] = [];
+
+    bodyLines.push(terminal.bold(terminal.green('WORKSPACE SCHEMATICS')));
```
Comment (vsavkin) [inline packages/workspace/src/command-line/workspace-schematic.ts:219]: We use helper functions from output.ts to format everything else. Maybe we use them here as well?

File: packages/workspace/src/command-line/workspace-schematic.ts [hunk 3]
```diff
@@ -323,29 +416,36 @@ async function executeSchematic(
...
-    alias: {
-      dryRun: ['d'],
-      listSchematics: ['l'],
-    },
+function parseSchematicOptions(
+  schematicArgs: string[],
+  schematicBooleanArgs: string[]
+) {
+  return yargsParser(schematicArgs, {
+    boolean: ['dryRun', 'interactive', 'verbose', ...schematicBooleanArgs],
```
Comment (vsavkin) [inline packages/workspace/src/command-line/workspace-schematic.ts:424]: Did you remove aliases?

**Review by vsavkin — APPROVED** [review]
A few minor comments

**Comment by vsavkin** [comment]
Thank you!

File: packages/workspace/src/command-line/workspace-schematic.ts (hunk 1 above)
Comment (Hotell) [reply packages/workspace/src/command-line/workspace-schematic.ts]: done, I'd suggest to enable proper eslint rules for that (sonarjs has such a rule)

File: packages/workspace/src/command-line/workspace-schematic.ts (hunk 2 above)
Comment (Hotell) [reply packages/workspace/src/command-line/workspace-schematic.ts:281]: I used output.ts where it was easy-to replace, some capabilities needs proper unification around codebase. I'd suggest to address those in another PR.

> for example
> - I just copied `printHelp` from `tao/src/command/generate.ts` which still uses deprecate angular-devkit terminal API
> - `listPluginCapabilities` from workspace/plugins uses bodyLines array "approach" etc

File: packages/workspace/src/command-line/workspace-schematic.ts (hunk 3 above)
Comment (Hotell) [reply packages/workspace/src/command-line/workspace-schematic.ts:489]: those aliases never worked for workspace-schematic (`-l`, `-d` )

Comment (vsavkin) [reply packages/workspace/src/command-line/workspace-schematic.ts:489]: Just checked. They work for me.

**Review by vsavkin — CHANGES_REQUESTED** [review]

File: packages/workspace/src/command-line/workspace-schematic.ts (hunk 3 above)
Comment (Hotell) [reply packages/workspace/src/command-line/workspace-schematic.ts:489]: - so `-l` doesn't work with latest nx 9 (which we are currently at)  
- `-d` works on both 9 and 10 

will add both back.

thanks

Comment (Hotell) [reply packages/workspace/src/command-line/workspace-schematic.ts:489]: 1. so I forgot that  `-l` is already covered within nx-commands.ts

image

2. `-d` is back
image

**Comment by vsavkin** [comment]
Unfortunately, the generation bit got reworked, and will be cleaned up in the next couple of months, so it was hard to get the whole change in. I managed to land a part of it (it is in master), and kept you as an author.

Sorry about that.

**Pull request closed without merging**